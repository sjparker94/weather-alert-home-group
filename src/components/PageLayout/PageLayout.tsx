import React from 'react';
import styled, { ThemeProvider } from 'styled-components';

import Meta from '../Meta/Meta';
import theme from '../../styles/theme';
import GlobalStyle from '../../styles/GlobalStyle';
import FavouritesSidebarStyles from '../FavouritesSidebar/FavouritesSidebarStyles';
import FavouritesSidebar from '../FavouritesSidebar/FavouritesSidebar';
import Header from '../Header/Header';
import { device } from '../../styles/breakpoint';

const StyledPage = styled.div`
    position: relative;
    min-height: 100vh;

    .main-page-layout-wrapper {
        height: 100vh;
        display: flex;
    }
    .main-page-content,
    ${FavouritesSidebarStyles} .favourites-wrapper {
        --scrollbarBG: ${props => props.theme.blueGrey};
        --thumbBG: ${props => props.theme.darkBlueGrey};
        overflow-y: auto;
        scrollbar-width: thin;
        scrollbar-color: var(--thumbBG) var(--scrollbarBG);
        &::-webkit-scrollbar {
            width: 12px;
            scrollbar-width: thin;
            scrollbar-color: var(--thumbBG) var(--scrollbarBG);
        }
        &::-webkit-scrollbar-track {
            background: var(--scrollbarBG);
        }
        &::-webkit-scrollbar-thumb {
            background-color: var(--thumbBG);
            border-radius: 6px;
            border: 3px solid var(--scrollbarBG);
        }
    }
    .main-page-content {
        flex: 1 1 100%;
    }

    ${FavouritesSidebarStyles} {
        flex: 1 0 600px;
        overflow-y: auto;
        .favourites-wrapper {
            --scrollbarBG: ${props => props.theme.lightBlueGrey};
            --thumbBG: ${props => props.theme.blueGrey};
        }
        @media ${device.laptopMMax} {
            flex: 1 0 300px;
        }
    }
`;

/** Shared layout across all pages */
const PageLayout: React.FC = ({ children }) => {
    return (
        <ThemeProvider theme={theme}>
            <StyledPage>
                <GlobalStyle />
                <Meta />
                <Header />
                <div className="main-page-layout-wrapper">
                    <main className="main-page-content">{children}</main>
                    <FavouritesSidebar />
                </div>
            </StyledPage>
        </ThemeProvider>
    );
};

export default PageLayout;
