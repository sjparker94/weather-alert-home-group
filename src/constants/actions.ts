export const GET_LOCATIONS_REQUEST = 'GET_LOCATIONS_REQUEST';
export const GET_LOCATIONS_SUCCESS = 'GET_LOCATIONS_SUCCESS';
export const GET_LOCATIONS_FAIL = 'GET_LOCATIONS_FAIL';

export const SEARCH_LOCATION_REQUEST = 'SEARCH_LOCATION_REQUEST';
export const SEARCH_LOCATION_SUCCESS = 'SEARCH_LOCATION_SUCCESS';
export const SEARCH_LOCATION_FAIL = 'SEARCH_LOCATION_FAIL';
export const SEARCH_LOCATION_CANCEL = 'SEARCH_LOCATION_CANCEL';

export const ADD_LOCATION = 'ADD_LOCATION';
export const REMOVE_LOCATION = 'REMOVE_LOCATION';

export const GET_LOCATION_FORECAST_REQUEST = 'GET_LOCATION_FORECAST_REQUEST';
export const GET_LOCATION_FORECAST_SUCCESS = 'GET_LOCATION_FORECAST_SUCCESS';
export const GET_LOCATION_FORECAST_FAIL = 'GET_LOCATION_FORECAST_FAIL';
