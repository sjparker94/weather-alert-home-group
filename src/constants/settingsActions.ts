export const TEMPERATURE_UNIT_TOGGLE = 'TEMPERATURE_UNIT_TOGGLE';
export const SPEED_UNIT_TOGGLE = 'SPEED_UNIT_TOGGLE';
