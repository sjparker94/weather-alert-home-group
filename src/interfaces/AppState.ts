import LocationsState from './LocationsState';
import SettingsState from './SettingsState';

interface AppState {
    locations: LocationsState;
    settings: SettingsState;
}

export default AppState;
