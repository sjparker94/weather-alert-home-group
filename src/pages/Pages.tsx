import React from 'react';
import { Route, Switch } from 'react-router-dom';
import HomePage from './HomePage';
import ErrorPage from './ErrorPage';
import LocationDetailPage from './LocationDetailPage';

const Pages: React.FC = () => {
    return (
        <Switch>
            <Route exact path="/" component={HomePage} />
            <Route exact path="/location/:id" component={LocationDetailPage} />
            <Route component={ErrorPage} />
        </Switch>
    );
};

export default Pages;
