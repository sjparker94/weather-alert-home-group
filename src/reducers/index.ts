import { combineReducers } from 'redux';
import locations from './locationsReducer';
import settings from './settingsReducer';

const rootReducer = combineReducers({
    locations,
    settings,
});

export default rootReducer;
