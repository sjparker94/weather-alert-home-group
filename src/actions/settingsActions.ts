import * as settingsActionTypes from '../constants/settingsActions';
import { Action } from 'redux';

export interface TemperatureUnitToggle extends Action {
    type: typeof settingsActionTypes.TEMPERATURE_UNIT_TOGGLE;
}
export interface SpeedUnitToggle extends Action {
    type: typeof settingsActionTypes.SPEED_UNIT_TOGGLE;
}

export const temperatureUnitToggle = (): TemperatureUnitToggle => {
    return {
        type: settingsActionTypes.TEMPERATURE_UNIT_TOGGLE,
    };
};
export const speedUnitToggle = (): SpeedUnitToggle => {
    return {
        type: settingsActionTypes.SPEED_UNIT_TOGGLE,
    };
};
