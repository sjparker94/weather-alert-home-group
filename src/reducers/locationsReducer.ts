import produce, { Draft, castDraft } from 'immer';
import LocationsState from '../interfaces/LocationsState';
import neverReached from '../utils/neverReached';
import { AllActions } from '../actions/allActions';
import * as LocationsActionTypes from '../constants/actions';
import { findByProp } from '../utils/arrayUtils';
import Location from '../interfaces/Location';

// type LocationsActionTypes = any;
export const locationsInitialState: LocationsState = {
    data: [],
    isInitialLoad: true,
    getLocations: {
        isPending: false,
        success: false,
    },
    searchLocation: {
        isPending: false,
        success: false,
        error: null,
        data: null,
    },
};

const locations = produce((draft: Draft<LocationsState>, action: AllActions) => {
    switch (action.type) {
        case LocationsActionTypes.GET_LOCATIONS_REQUEST:
            draft.getLocations.isPending = true;
            break;
        case LocationsActionTypes.GET_LOCATIONS_SUCCESS:
            draft.data = castDraft(action.payload);
            draft.isInitialLoad = false;
            break;
        case LocationsActionTypes.SEARCH_LOCATION_REQUEST:
            draft.searchLocation.isPending = true;
            draft.searchLocation.data = null;
            draft.searchLocation.error = null;
            break;
        case LocationsActionTypes.SEARCH_LOCATION_SUCCESS:
            draft.searchLocation.isPending = false;
            draft.searchLocation.error = null;
            draft.searchLocation.data = action.payload;
            break;
        case LocationsActionTypes.SEARCH_LOCATION_CANCEL:
            draft.searchLocation.data = null;
            draft.searchLocation.error = null;
            break;
        case LocationsActionTypes.SEARCH_LOCATION_FAIL:
            draft.searchLocation.isPending = false;
            draft.searchLocation.error = action.payload;
            break;
        case LocationsActionTypes.ADD_LOCATION:
            // See if the location is already in favourites
            const currentIdx = draft.data.findIndex(findByProp<Location>('id', action.payload.id));

            if (currentIdx === -1) {
                draft.data.push(action.payload);
            } else {
                // Update existing location with the latest data fetched
                draft.data[currentIdx] = action.payload;
            }
            draft.searchLocation.data = null;
            break;
        default:
            neverReached(action);
    }
}, locationsInitialState);

export default locations;
