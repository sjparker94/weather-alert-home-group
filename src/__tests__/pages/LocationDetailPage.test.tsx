import React from 'react';
import { Router, Route } from 'react-router-dom';
import axios from 'axios';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom/extend-expect';
import { createMemoryHistory } from 'history';

// @ts-ignore
import { render, screen, fireEvent, waitFor, fakeLocation } from '../../utils/testUtils';
import Pages from '../../pages/Pages';
import AppState from '../../interfaces/AppState';
import { locationsInitialState } from '../../reducers/locationsReducer';
import { settingsInitialState } from '../../reducers/settingsReducer';
import theme from '../../styles/theme';
import PageLayout from '../../components/PageLayout/PageLayout';
import { speedUnitToggle } from '../../actions/settingsActions';

jest.mock('axios');

const mockLocation = fakeLocation();
const mockLocation2 = fakeLocation({
    id: 123,
    name: 'Newcastle upon Tyne',
});
mockLocation2.wind.deg = 146;
const mockLocation3 = fakeLocation({
    id: 456,
    name: 'Somewhere else',
});

const mockInitalReduxStateWithData: AppState = {
    locations: {
        ...locationsInitialState,
        isInitialLoad: false,
        data: [mockLocation, mockLocation2],
    },
    settings: {
        ...settingsInitialState,
    },
};

// Need to render full app with router to test the toggle of the settings
describe('<LocationDetailPage />', () => {
    it('renders with the correct location and matches snapshot', async () => {
        const history = createMemoryHistory({
            // id matches one of the items in the mock redux store
            initialEntries: ['/location/123'],
        });
        const { container } = render(
            <Router history={history}>
                <Pages />
            </Router>,
            mockInitalReduxStateWithData
        );

        // Check that the correct city is showing based on the id used in the url
        expect(screen.getByText(/Newcastle upon Tyne/i)).toBeInTheDocument();

        expect(container).toMatchSnapshot();
    });

    it('renders with the correct location and is displaying the data correctly', async () => {
        const history = createMemoryHistory({
            // id matches one of the items in the mock redux store
            initialEntries: ['/location/123'],
        });
        render(
            <Router history={history}>
                <Pages />
            </Router>,
            mockInitalReduxStateWithData
        );

        const expectedWindDirectionCssValue = `${mockLocation2.wind.deg - 180}deg`;
        const expectedWindMph = '6';
        const expectedWindSpeedColor = theme.green;
        const expectedSpeedUnit = 'mph';
        const expectedTempUnit = '°C';
        const expectedSpeedText = 'Gentle';
        const expectedDirectionText = 'South Easterly';

        // Get elements
        const windDirectionIcon = screen.getByTestId('wind-direction-icon');
        const mainWindSpeedHeading = screen.getByTestId('main-wind-speed-title');
        const windSpeedTextValue = screen.getByTestId('wind-speed-text-value');
        const windDirectionTextValue = screen.getByTestId('wind-direction-text-value');
        const speedUnitEls = screen.getAllByTestId('speed-units');
        const tempUnitEls = screen.getAllByTestId('temp-units');

        // Check wind direction icon is showing and that the correct rotate
        expect(windDirectionIcon).toHaveStyle(
            `transform: rotate(${expectedWindDirectionCssValue});`
        );
        // Check the value of wind is showing
        expect(mainWindSpeedHeading).toHaveTextContent(expectedWindMph);
        // Check that the variable color scheme is displaying the right color value
        expect(mainWindSpeedHeading).toHaveStyle(`color: ${expectedWindSpeedColor}`);
        // check that the textual value is correct
        expect(windSpeedTextValue).toHaveTextContent(expectedSpeedText);
        // Check that the textual direction is correct
        expect(windDirectionTextValue).toHaveTextContent(expectedDirectionText);

        // loop all of the speed & temp unit elements and check they are set the to initial redux store value
        speedUnitEls.forEach(el => {
            expect(el).toHaveTextContent(expectedSpeedUnit);
        });
        tempUnitEls.forEach(el => {
            expect(el).toHaveTextContent(expectedTempUnit);
        });
    });

    // rendering with the PageLayout so we can use the
    it('show correct location and toggles units and values correctly', async () => {
        const history = createMemoryHistory({
            // id matches one of the items in the mock redux store
            initialEntries: ['/location/123'],
        });
        render(
            <Router history={history}>
                <PageLayout>
                    <Route component={Pages} />
                </PageLayout>
            </Router>,
            mockInitalReduxStateWithData
        );

        // All of the values that can change via settings
        const expectedChangedWindValue = '9';
        const expectedChangedSpeedUnit = 'km/h';
        const expectedChangedTempUnit = '°F';

        // Get elements
        const mainWindSpeedHeading = screen.getByTestId('main-wind-speed-title');
        const speedUnitEls = screen.getAllByTestId('speed-units');
        const tempUnitEls = screen.getAllByTestId('temp-units');

        const settingsButton = screen.getByRole('button', {
            name: 'Open Settings',
        });

        // fire events to toggle settings menu open
        fireEvent.click(settingsButton);

        const tempToggle = screen.getByRole('checkbox', {
            name: 'Temperature unit',
        });
        const speedToggle = screen.getByRole('checkbox', {
            name: 'Wind speed unit',
        });

        // fire events to toggle the units
        fireEvent.click(tempToggle);
        fireEvent.click(speedToggle);

        // Check the value of wind has changed to correct km/h value
        expect(mainWindSpeedHeading).toHaveTextContent(expectedChangedWindValue);

        // loop all of the speed & temp unit elements and check they have switched
        speedUnitEls.forEach(el => {
            expect(el).toHaveTextContent(expectedChangedSpeedUnit);
        });
        tempUnitEls.forEach(el => {
            expect(el).toHaveTextContent(expectedChangedTempUnit);
        });
    });

    it('redirects to the error page when location is deleted and you are on that page', () => {
        const history = createMemoryHistory({
            // id matches one of the items in the mock redux store
            initialEntries: ['/location/123'],
        });
        render(
            <Router history={history}>
                <PageLayout>
                    <Route component={Pages} />
                </PageLayout>
            </Router>,
            mockInitalReduxStateWithData
        );

        const sidebarDeleteEl = screen.getByRole('button', {
            name: 'Remove Newcastle upon Tyne from favourites',
        });
        fireEvent.click(sidebarDeleteEl);

        expect(history.location.pathname).toBe('/404');
    });
});
