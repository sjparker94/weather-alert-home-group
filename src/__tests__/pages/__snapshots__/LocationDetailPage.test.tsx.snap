// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LocationDetailPage /> renders with the correct location and matches snapshot 1`] = `
<div>
  <section
    class="sc-fzpans gdQUzA"
  >
    <section
      class="sc-fzozJi eRVneo"
    >
      <div
        class="page-title-left-right-wrapper"
      >
        <div
          class="page-title-content"
        >
          <h1>
            Newcastle upon Tyne
          </h1>
          <h3>
            United Kingdom of Great Britain and Northern Ireland
          </h3>
        </div>
        <div
          class="page-title-extra"
        >
          <p
            class="min-max-temp"
          >
            <svg
              class="up-icon"
              fill="#2ECC40"
              height="24"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M7 14l5-5 5 5z"
              />
            </svg>
            <span>
              24
              <sup>
                째
              </sup>
            </span>
            <svg
              fill="#E31B3A"
              height="24"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <span>
              22
              <sup>
                째
              </sup>
            </span>
          </p>
          <h3>
            <img
              alt="Clouds"
              src="http://openweathermap.org/img/wn/03d.png"
            />
            23
            <sup
              data-testid="temp-units"
            >
              째C
            </sup>
          </h3>
          <p
            class="feels-like"
          >
            Feels like 
            20
            <sup>
              째
            </sup>
          </p>
        </div>
      </div>
    </section>
    <div
      class="content-wrapper"
    >
      <div
        class="sc-AxheI kwSqZm"
      >
        <div
          class="sc-AxgMl cVkjKg"
        >
          <div
            class="icon-wrapper"
          >
            <svg
              height="34px"
              version="1.1"
              viewBox="0 0 32 32"
              width="34px"
              xmlns="http://www.w3.org/2000/svg"
              xmlns:xlink="http://www.w3.org/1999/xlink"
            >
              <g
                fill="none"
                fill-rule="evenodd"
                stroke="none"
                stroke-width="1"
              >
                <g
                  fill="#718096"
                  id="icon-43-wind"
                >
                  <path
                    d="M30,14 C30,11.790861 28.2046438,10 26,10 C23.790861,10 22,11.7894101 22,14 L23,14 C23,12.3431458 24.3465171,11 26,11 C27.6568542,11 29,12.3465171 29,14 C29,15.6568542 27.6568766,17 25.9920699,17 L3,17 L3,18 L26.0015293,18 C28.2098237,18 30,16.2046438 30,14 L30,14 Z M21,13 C21,11.3431458 19.6534829,10 18,10 C16.3431458,10 15,11.3420576 15,12.991394 L15,13 L16,13 C16,11.8954305 16.8877296,11 18,11 C19.1045695,11 20,11.8877296 20,13 C20,14.1045695 19.1029399,15 17.9941413,15 L6,15 L6,16 L18.0005775,16 C19.6571128,16 21,14.6534829 21,13 L21,13 Z M25,21 C25,22.1045695 24.1122704,23 23,23 L23,23 C21.8954305,23 21,22.1052949 21,21.0057373 L21,21 L22,21 C22,21.5522847 22.4438648,22 23,22 L23,22 C23.5522847,22 24,21.5561352 24,21 L24,21 C24,20.4477153 23.5510798,20 22.992516,20 L9,20 L9,19 L23.0059397,19 C24.1072288,19 25,19.8877296 25,21 L25,21 L25,21 Z"
                    id="wind"
                  />
                </g>
              </g>
            </svg>
          </div>
          <h4>
            Wind Conditions
          </h4>
        </div>
        <div
          class="sc-fznyAO yloHq"
        >
          <div
            class="wind-details-speeds"
          >
            <div
              class="value-wrapper"
            >
              <h2
                data-testid="main-wind-speed-title"
              >
                6
              </h2>
              <span
                class="units"
                data-testid="speed-units"
              >
                mph
              </span>
            </div>
            <div
              class="direction-wrapper"
            >
              <svg
                class="wind-direction-icon"
                data-testid="wind-direction-icon"
                fill="#718096"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M12 2L4.5 20.29l.71.71L12 18l6.79 3 .71-.71z"
                />
              </svg>
            </div>
          </div>
          <div
            class="wind-details-content"
          >
            <h3
              data-testid="wind-speed-text-value"
            >
              Gentle
            </h3>
            <p
              data-testid="wind-direction-text-value"
            >
              South Easterly
            </p>
          </div>
          <div
            class="upcoming-wind-details"
          >
            <h4>
              Upcoming 24 Hours
            </h4>
            <div
              class="loader-wrapper"
            >
              <div
                aria-busy="true"
                class="sc-fznKkj dijmJu"
              >
                <div
                  class="dot1"
                />
                <div
                  class="dot2"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <h2>
        Current Details
      </h2>
      <div
        class="cols-wrapper"
      >
        <div
          class="sc-AxheI sc-fznZeY lgdgpQ"
        >
          <h4>
            Summary
          </h4>
          <p
            class="value"
          >
            Clouds
          </p>
          <p
            class="description"
          >
            Scattered clouds
          </p>
        </div>
        <div
          class="sc-AxheI sc-fznZeY lgdgpQ"
        >
          <h4>
            Humidity
          </h4>
          <p
            class="value"
          >
            31
          </p>
          <p
            class="description"
          >
            %
          </p>
        </div>
        <div
          class="sc-AxheI sc-fznZeY lgdgpQ"
        >
          <h4>
            Visibility
          </h4>
          <p
            class="value"
          >
            6
          </p>
          <p
            class="description"
          >
            miles
          </p>
        </div>
      </div>
      <div
        class="loader-wrapper"
      >
        <div
          aria-busy="true"
          class="sc-fznKkj dijmJu"
        >
          <div
            class="dot1"
          />
          <div
            class="dot2"
          />
        </div>
      </div>
    </div>
  </section>
</div>
`;
